src/cmap/commands.ts

```diff
@@ -709,3 +715,54 @@ export class BinMsg {
     return { utf8: { writeErrors: false } };
   }
 }
+
+const MESSAGE_HEADER_SIZE = 16;
+const COMPRESSION_DETAILS_SIZE = 9; // originalOpcode + uncompressedSize, compressorID
+
+/**
+ * @internal
+ *
+ * An OP_COMPRESSED request wraps either an OP_QUERY or OP_MSG message.
+ */
+export class OpCompressedRequest {
+  constructor(
+    private command: WriteProtocolMessageType,
+    private options: { zlibCompressionLevel: number; agreedCompressor: CompressorName }
+  ) {}
+
+  // Return whether a command contains an uncompressible command term
+  // Will return true if command contains no uncompressible command terms
+  static canCompress(command: WriteProtocolMessageType) {
+    const commandDoc = command instanceof OpMsgRequest ? command.command : command.query;
+    const commandName = Object.keys(commandDoc)[0];
```

Comment:
enumerating all keys and checking the 0th would be no different from checking for the key existing in the commandDoc right? Ex. a `find` can't have `saslContinue` as a top-level key.

Reply:
Can we say that confidently for all commands? 🤔 

While maybe a worthwhile improvement, I'd rather file a follow up ticket for that.  Or even better, just pass the operation down into the connection layer and use an `INCOMPRESSIBLE` aspect on the operation here instead

Reply:
I am fairly (99.99999%) certain we can but fair enough if we want to file something and leave a TODO.

ismaster, saslStart, saslContinue, are the only commands we make in the driver but they are not operations. The rest can be run by someone else via runCommand so the decision does have to be made based on the command.

Reply:
I'd say a follow up is fine but it seems the server expects the first key to always be the command name and if another command name is in the top level the server will error:

```shell
Enterprise test> db.runCommand({ find: 'test', saslContinue:  {} })
MongoServerError: BSON field 'FindCommandRequest.saslContinue' is an unknown field.
Enterprise test> 
```